### PR TITLE
[WIP] Refresh token support

### PR DESCRIPTION
**Pull request by IsNull**
This is the continuation of the PR #5 and aims at supporting refresh-tokens.

### CODE REVIEW

File: src/auth/refresh/ondemand-token-refresh.service.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    public ensureRefreshment() {
+        if (!this.config.refreshTokenUrl || this.subscription) {
+            return;
+        }
+
+        this.logger.debug('Starting on demand token refreshment');
+        const tokenTtl = this.jwtHelper.getTokenExpirationDate(this.tokenService.getAccessToken()).getTime() -
+            (new Date()).getTime();
+        if (tokenTtl < 5000) {
```
Comment (GNSSYSTEMS-tg) [inline src/auth/refresh/ondemand-token-refresh.service.ts]: Now that I look at it, this should be static. Same as below. IntelliJ would call it "magic number".

Another thought: is 5 seconds sufficient?

Comment (IsNull) [reply src/auth/refresh/ondemand-token-refresh.service.ts]: Usually 60 Seconds before expiration is used but I consider to make this configurable aswell.

**Comment by GNSSYSTEMS-tg** [comment]
Btw. tagging of your versions in git would be nice. So everyone can jump between 1.1 and 2.0 for instance.

**Comment by IsNull** [comment]
@GNSSYSTEMS-tg I've added the tags. Note however that we do not plan to support older branches as long we have not reached a stable version.

I have now basically completed the 2.x release, and you can try the `2.0.0-beta-7` version. I've tested it in some projects already, however the refresh token support has not been tested in production.

**Comment by GNSSYSTEMS-tg** [comment]
Don't worry, neither do I, but thanks for the tags. 

I'll have a peak at the current beta.

**Comment by GNSSYSTEMS-tg** [comment]
Ok, so the `refreshTokenUrl` is missing in the config ;) It does exist within the `AuthConfig` but it's never initialized nor accessible within `JwtAuthConfig`.

File: src/auth/auth.config.ts [hunk 2]
```diff
@@ -36,17 +36,16 @@ export class JwtAuthConfig {
...
+     * grant_type: refresh_token    -> Get a JWT for the refresh token
+     *
+     * Should be an URL like '/oauth/token'
      */
     obtainTokenUrl?: string;
 
     /**
-     * Provides URL for refreshing JWT before expiration.
-     */
-    refreshTokenUrl?: string;
```
Comment (GNSSYSTEMS-tg) [inline src/auth/auth.config.ts]: My API works like this:
`/api/token/obtain`: obtain token
`/api/token/refresh`: refresh token

Why souldn't this be supported?

**Comment by IsNull** [comment]
Ah I see. The OAuth standard would be to have different grant_types in the same token url. I can optionally support configuring a separate refresh url and by default its the obtain token url.

**Comment by IsNull** [comment]
Beta `2.0.0-beta.8` is released which supports a customisable token refresh url.

**Comment by GNSSYSTEMS-tg** [comment]
There is an issue with refreshing after the first refresh:

~~~
2017-09-21T14:51:51Z [DEBUG] Obtaining JWT for OAuthGrant.grant_type: password
2017-09-21T14:51:51Z [DEBUG] Token time to live: 69718
>>> logged in
2017-09-21T14:52:01Z [DEBUG] Refreshing JWT.
2017-09-21T14:52:01Z [DEBUG] Obtaining JWT for OAuthGrant.grant_type: refresh_token
~~~

And then the logging of the new token ttl is missing, so I suppose the new interval is not created.

For some reason, the observable of the token container is not triggered after refresh...

**Comment by GNSSYSTEMS-tg** [comment]
Looks fine to me.

**Comment by GNSSYSTEMS-tg** [comment]
Crap, there's one more thing...

**Pull request merged**